### PR TITLE
Rl state

### PR DESCRIPTION
**Pull request by nicelhc13**
Add base mechanism for RL (state creation, A2C, buffering)

### CODE REVIEW

**Comment by nicelhc13** [comment]
In this PR, I also added equations that we discussed. I implemented them based on my understanding. Please check it they are aligned with your understanding. Please check simple.py and oracles.py.

Thank you

File: src/task4feedback/simulator/rl/models/a2c.py [hunk 1]
```diff
@@ -48,106 +67,103 @@ def compute_returns(self, next_value, rewards):
...
-            self.lst_rewards = []
-
+            actor_loss = -(cat_log_probs * advantage.detach()).mean()
+            #critic_loss = advantage.pow(2).mean()
+            critic_loss = 1 * F.mse_loss(cat_values.unsqueeze(-1), returns.unsqueeze(-1))
+            loss = actor_loss + 0.5 * critic_loss - 0.001 * self.entropy_sum
             self.optimizer.zero_grad()
             loss.backward()
+            # torchviz.make_dot(loss, params=dict(self.a2c_model.named_parameters())).render("attacehd", format="png")
             for param in self.a2c_model.parameters():
```
[the remark this answers is not in the document]
Comment (wlruys) [reply src/task4feedback/simulator/rl/models/a2c.py:144]: Clip grad norm is safer to preserve direction, right? Or am I missing something?

File: src/task4feedback/simulator/rl/models/env.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+      self.task_property_len = 4 + self.num_task_types + num_devices
+      self.device_utilization_state_offset = self.task_property_offset + \
+                                             self.task_property_len
+      self.device_state_len = 3 * num_devices
+      self.state_dim = self.device_utilization_state_offset + \
+                       self.device_state_len
+      self.out_dim = num_devices
+      self.num_devices = num_devices
+      if logger.ENABLE_LOGGING:
+          logger.runtime.info(f"RL state dimension: {self.state_dim}.")
```
Comment (wlruys) [inline src/task4feedback/simulator/rl/models/env.py:64]: Recommendation: Please use a new logger type (logger.reinforcement?) so its easy to debug/filter runtime issues (launching/mapping/reserving) separately from network training and testing.

File: src/task4feedback/simulator/rl/models/oracles.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+class OraclePolicy(ABC):
+
+  def get_action(self, system_state: "SystemState"):
+      NotImplementedError()
+
+
+class LoadbalancingPolicy(ABC):
+
+
+  def get_action(self, system_state: "SystemState"):
```
Comment (wlruys) [inline src/task4feedback/simulator/rl/models/oracles.py:16]: Eventually for safety this should somehow:
(1) Filter for devices that can't run the current task; or 
(2) Be passed a device list that meets the current tasks requirements and makes a choice over that list.

File: src/task4feedback/simulator/rl/models/oracles.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                      total_workload_for_next_norm += action_probs[device.device_id].item()
+              else:
+                  # If all workload is 0 (so all devices are idle), assign 1/(#devices) as its action prob.
+                  action_probs[device.device_id] = 1 / float(len(devicemap) - 1)
+          if total_workload > 0 and need_further_norm:
+              for device in devicemap:
+                  if device.architecture == Architecture.CPU:
+                      continue;
+
+                  action_probs[device.device_id] = action_probs[device.device_id].item() / total_workload_for_next_norm
```
Comment (wlruys) [inline src/task4feedback/simulator/rl/models/oracles.py:60]: This control flow seems much more confusing than it needs to be. 
Would the performance impact of summing and renormalizing every step be significant enough to justify this?

Can this be refactored to just have:
A[i] = W[i] 

if SUM(A):
     A[:] = A[i]/SUM(A)
else:
     A[:] = 1 / LEN(A)

File: src/task4feedback/simulator/rl/models/simple.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        print("usq concat z:", concat_z.unsqueeze(-1))
+        with torch.no_grad():
+            print("crossentropy:", F.cross_entropy(concat_p, concat_pi))
+            print("crossentropy mean:", F.cross_entropy(concat_p, concat_pi).mean())
+            # print("usq crossentropy:", F.cross_entropy(concat_p.unsqueeze(-1), concat_pi.unsqueeze(-1)))
+            print("usq mse loss:", F.mse_loss(concat_v.unsqueeze(-1), concat_z.unsqueeze(-1)))
+            print("mse loss:", F.mse_loss(concat_v, concat_z))
+        """
+
+        loss = -F.cross_entropy(concat_p, concat_pi, reduction='sum') + \
```
Comment (wlruys) [inline src/task4feedback/simulator/rl/models/simple.py:155]: No regularization?

File: src/task4feedback/simulator/rl/networks/a2c_fcn.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import torch
+import torch.nn
+import torch.nn.functional as F
+from torch.nn import Linear
+from torch_geometric.nn import GCNConv
+
+
+class A2CNetworkNoGCN(torch.nn.Module):
+
+    def __init__(self, in_dim: int, out_dim: int):
```
Comment (wlruys) [inline src/task4feedback/simulator/rl/networks/a2c_fcn.py:10]: Is it okay that the Actor and Critic are separate networks? 
"the literature" seems divided on this, with arguments both ways haha 
I don't have the experience to recommend either just curious to note it.

File: src/task4feedback/simulator/schedulers/parla/architecture.py [hunk 7]
```diff
@@ -671,4 +824,9 @@ def complete(self, scheduler_state: SystemState) -> bool:
                     complete_flag and self.launchable_tasks[device][task_type].empty()
                 )
 
+        reward = scheduler_state.target_exec_time / convert_to_float(
+            scheduler_state.time.scale_to("ms"))
+        print("Total execution time:", scheduler_state.time, " reward:", reward)
+        self.rl_mapper.optimize_model(reward)
```
Comment (wlruys) [inline src/task4feedback/simulator/schedulers/parla/architecture.py:830]: Just to check, this is the only time optimize_model is called right? Between episodes?

File: src/task4feedback/simulator/rl/models/simple.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    def select_device(self, state: torch.tensor, oracle):
+        """
+        Select a device from pi.
+        If a specified state has not been visited, select a device from a neural
+        network.
+        """
+
+        state_tuple = tuple(state.tolist())
+        with torch.no_grad():
+            if state_tuple in self.pi:
```
Comment (wlruys) [inline src/task4feedback/simulator/rl/models/simple.py:69]: I don't understand this

Comment (wlruys) [reply src/task4feedback/simulator/rl/models/simple.py:69]: Why does the pi(S) -> action need to be stored for all states?

File: src/task4feedback/simulator/schedulers/parla/architecture.py [hunk 9]
```diff
@@ -633,9 +780,17 @@ def complete_task(
         self, scheduler_state: SystemState, event: TaskCompleted
     ) -> List[EventPair]:
         objects = scheduler_state.objects
+
         task = objects.get_task(event.task)
         current_time = scheduler_state.time
 
+        next_events: List[EventPair] = []
+        self.success_count += 1
```
[the remark this answers is not in the document]
Comment (wlruys) [reply src/task4feedback/simulator/schedulers/parla/architecture.py:788]: ..I don't think it matters because a "task completion failure" means the program crashes anyway. My preference is to only increment "success" after it's been completed though

File: src/task4feedback/simulator/rl/models/simple.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    def update_pi(self, state: torch.tensor, state_tuple: Tuple[float], oracle):
+        """ Update target policy, pi.
+        """
+
+        # Buffer probabilities from network
+        actions, v = self.network(NetworkInput(state, False, None, None))
+        actions = F.softmax(actions, dim=0)
+        # print("oracle:", oracle, " softmax dnn:", actions)
+        # Calculate loss
+        self.pi[state_tuple] = (1 - self.ld) * actions + self.ld * oracle.to(self.device)
```
Comment (wlruys) [inline src/task4feedback/simulator/rl/models/simple.py:94]: I think ld should be random. 
We discussed it as a bernoulli rv to toggle between either at each step

Comment (wlruys) [reply src/task4feedback/simulator/rl/models/simple.py:94]: Although (if training is fast), it is maybe worth seeing if this converges even with a static ld.

File: src/task4feedback/simulator/schedulers/parla/architecture.py [hunk 11]
```diff
@@ -35,8 +42,117 @@ def chose_random_placement(task: SimulatedTask) -> Tuple[Device, ...]:
...
 def map_task(
+    task: SimulatedTask, scheduler_state: ParlaState,
+    rl_mapper: RLModel, rl_env: RLBaseEnvironment, oracle: LoadbalancingPolicy,
+    verbose: bool = False
+) -> Optional[Tuple[Device, ...]]:
+    """
+    Invoke a specified task mapper
+    """
+    exec_mode = scheduler_state.exec_mode
+    if exec_mode == ExecutionMode.RL_TRAINING or exec_mode == ExecutionMode.RL_TESTING or \
```
[the remark this answers is not in the document]
Comment (wlruys) [reply src/task4feedback/simulator/schedulers/parla/architecture.py:53]: My reasoning is that ideally you could switch implementation in model files and the runner script / config files without having to reflect any changes in architecture once the interface is defined.

File: src/task4feedback/simulator/rl/models/a2c.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/a2c.py:144]: this is for gradient exploding. I used this for the past model since I saw that due to high state spaces. But it might not be necessary in here since we all normalize states.

File: src/task4feedback/simulator/rl/models/simple.py (hunk 8 above)
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/simple.py:69]: This is also the thing that I confused (So I tried to ask this question to you in the slack).
pi(S) is a mapping bewteen S to action probs, correct(so, like a dictionary)? Then how can I avoid to store states?
or do you mean that it should be a function with an equation? what is that equation?

Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/simple.py:69]: also this pi(S) should be updated for each iteration (so, at the end of an iteration k, pi_{k-1}(S), but this will be updated based on oracle and f(theta) at an iteration k, to pi_{k}(S)

Comment (wlruys) [reply src/task4feedback/simulator/rl/models/simple.py:69]: You only need to store the pi associated with a given action choice, iiuc

`
mapping_function():
        # get current state S
        # compute pi, p, v 
        # use pi to decide mapping a
        buffer.add( (S, pi, p, v, a) ) 
        
optimize_network():
          # use buffer to compute loss 
          # backprop & step
          # clear buffer
`

Where:
 S is a feature vector
 pi is a vector of length(# devices)
 p is a vector of length(# devices)
 v is a scalar
 a is a scalar

Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/simple.py:69]: I don't get it. Sorry. what do you mean by "compute pi"? 
Then, do you mean that we have an equation to compute pi?
What is that?
lambda*f + (lambda-1)*oracle is an equation to "update" pi, isn't it?

Comment (wlruys) [reply src/task4feedback/simulator/rl/models/simple.py:69]: Given a state S and a network function f  `lambda*f(S) + (lambda-1)*oracle(S)` is a function that computes pi. 
I am not sure what you mean by update.

Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/simple.py:69]: ok. that's my misunderstanding. in the jamboard, this was introduced as "update" function. in alphago, it has an update function from bottom to top with its simulated `v`. So there is no update but only decision from oracle. This is good and now I think I understood it. My current code is then incorrect. Its updating pi function also should be removed.

File: src/task4feedback/simulator/schedulers/parla/state.py [hunk 12]
```diff
@@ -1091,3 +1091,236 @@ def completion_stats(self, task: SimulatedTask):
 
                 data.stats.move_time += task.duration
                 data.stats.move_count += 1
+
+
+@SchedulerOptions.register_state("rl")
+@dataclass(slots=True)
+class RLState(SystemState):
```
[the remark this answers is not in the document]
Comment (nicelhc13) [reply src/task4feedback/simulator/schedulers/parla/state.py:1098]: I think this might be from my misunderstanding on the decorator?
If I inherit "ParlaState", is this class registered as "parla" or "rl"?
If the decorator doesn't care about its base class, and this class is registered as only "rl", then I will modify that.

[the remark this answers is not in the document]
Comment (wlruys) [reply src/task4feedback/simulator/schedulers/parla/state.py:1098]: It'll work for this (architecture/state) decorator as long as it's in the same class hierarchy since it's building a string to class type dictionary. 

You're right that some of the other decorators like the `register_graph` would care about the type of the arguments in the signature and might have issues as written with this kind of inheritance. But the state/arch should be fine.

File: src/task4feedback/simulator/schedulers/parla/architecture.py (hunk 7 above)
Comment (nicelhc13) [reply src/task4feedback/simulator/schedulers/parla/architecture.py:830]: Right!

File: src/task4feedback/simulator/rl/networks/a2c_fcn.py (hunk 6 above)
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/networks/a2c_fcn.py:10]: I think this should be renamed to something new.. not A2C. A2C is not about a specific network, but more RL architecture. For example, we consider state-action probs and state value, but this uses return which is (state-action probs - state value). A2C needs to use 2 networks, actor and critic. But it doesn't matter what network they internally use. Sorry I just didn't know how to name it :-\

File: src/task4feedback/simulator/rl/models/oracles.py (hunk 4 above)
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/oracles.py:60]: I agree but we need one more condition in case that there are both idle and non-idle devices. In this case, the idle device will get prob of 0. However, this should have the biggest prob over other non-idle devices. The code can be simplified further. but I am not sure if we can avoid one additional loop to see if there is idle device.

File: src/task4feedback/simulator/rl/models/simple.py (hunk 5 above)
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/simple.py:155]: If I understood correctly, if I set an weight decay parameter to the optimizer, L2 penalty will be added to the weight. I will add it.

File: src/task4feedback/simulator/rl/models/oracles.py [hunk 13]
```diff
@@ -23,40 +23,30 @@ def get_action(self, system_state: "SystemState"):
...
-                  need_further_norm = True
-              total_workload += workload
-              action_probs[device.device_id] = workload
+          for device in devicemap:
+              # Ignore CPU device
+              if device.architecture == Architecture.CPU:
+                  continue
 
-          total_workload_for_next_norm = 0
+              business = max_worload - system_state.perdev_active_workload[device]
```
[the remark this answers is not in the document]
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/oracles.py]: replaced it with your awesome equatin.

File: src/task4feedback/simulator/rl/models/simple.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        If a specified state has not been visited, select a device from a neural
+        network.
+        """
+
+        state_tuple = tuple(state.tolist())
+        actions, v = self.network(NetworkInput(state, False, None, None))
+        f_action_probs = F.softmax(actions, dim=0)
+        # rnd_ld = random.choice([1, 0])
+        # rnd_ld = 1
+        rnd_ld = random.uniform(0, 1)
```
[the remark this answers is not in the document]
Comment (wlruys) [reply src/task4feedback/simulator/rl/models/simple.py:65]: otherwise this "learns to make good decisions" given the teacher heuristic, and doesn't need to be a good policy on its own

[the remark this answers is not in the document]
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/simple.py:65]: That makes sense to me but any random float number b/w [0, 1] is not wrong as long as f contributes at some degree. My strategy is to test multiple options. I will test lambda closing to 1.

**Comment by wlruys** [comment]
https://github.com/wlruys/task4feedback/blob/b709e2a9c2f4f722c87e6f21368eebbf808564a5/src/task4feedback/simulator/interface.py#L12

File: src/task4feedback/simulator/schedulers/parla/architecture.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (nicelhc13) [reply src/task4feedback/simulator/schedulers/parla/architecture.py]: Why don't we just use ParlaState or RLState to dispatch proper mappers? I already made the states to have the mappers but also I think adding them on either the SimulatorConfig or *State looks similar IMO. (Until we have a config or an operation class that manages specifically each phase operations)

File: src/task4feedback/simulator/rl/models/a2c.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicelhc13) [reply src/task4feedback/simulator/rl/models/a2c.py:144]: i didn't put much energy to this. (https://jamesmccaffrey.wordpress.com/2022/10/17/the-difference-between-pytorch-clip_grad_value_-and-clip_grad_norm_-functions/) just used simple and easy to use as the article says. 

> Do most gradients fall in (-1,1)? 
Input gradients were not within (-1, 1). In the past simulator and model, gradients were aggressive and probs was exploded. I don't have concrete numbers now. But it seems like it is worth to explore in the new model + simulator.

File: src/task4feedback/simulator/utility.py [hunk 15]
```diff
@@ -115,6 +111,220 @@ def get_heft_rank(task):
...
+                intersected_data = []
+                intersected_data_size = 0
+                for sd in src_data:  # SimulatedData
+                    # If the data is already planned to be materialized on this device,
+                    # skip it.
+                    if sd.id in data_cache[current_device]:
+                        continue
+
+                    for dd in all_dep_data:  # DataID
+                        if sd.id == dd:
```
Comment (wlruys) [inline src/task4feedback/simulator/utility.py:201]: I think this loops still over counts data that is shared on multiple incoming edges. 
For example, a task that writes to block A that depends on two other tasks that both read block A

Comment (nicelhc13) [reply src/task4feedback/simulator/utility.py:205]: I am aware of it but this consideration would require more policeis in HEFT. 
For example, which dependency should we consider?
We should consider which dependency would be the best option for this task. 
One dependency might be finished earlier than others. But due to its placement, data transfer time would be more expensive. We need to consider such factors and I think this would make over complex the policy. So for now, I just skipped it.
